closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -1855,3 +1955,549 @@ int OSSL102_RSA_set0_crt_params(RSA *r2, BIGNUM *dmp1, BIGNUM *dmq1, BIGNUM *iqm
+        (*env)->ReleasePrimitiveArrayCritical(env, y, nativeY, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+        (*OSSL_BN_CTX_free)(ctx);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    group = (*OSSL_EC_GROUP_new_curve_GFp)(pBN, aBN, bBN, ctx);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, a, nativeA, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, b, nativeB, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, p, nativeP, JNI_ABORT);
+
+    generator = (*OSSL_EC_POINT_new)(group);
+
+    if ((NULL == group) || (NULL == generator)) {
+        (*env)->ReleasePrimitiveArrayCritical(env, x, nativeX, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, y, nativeY, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_BN_CTX_free)(ctx);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    ret = (*OSSL_EC_POINT_set_affine_coordinates_GFp)(group, generator, xBN, yBN, ctx);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, x, nativeX, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, y, nativeY, JNI_ABORT);
+    (*OSSL_BN_CTX_free)(ctx);
+
+    if (0 == ret) {
+        (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+        (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    ret = (*OSSL_EC_GROUP_set_generator)(group, generator, nBN, hBN);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, n, nativeN, JNI_ABORT);
+    (*env)->ReleasePrimitiveArrayCritical(env, h, nativeH, JNI_ABORT);
+
+    if (0 == ret) {
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    ret = (*OSSL_EC_KEY_set_group)(key, group);
+
+    if (0 == ret) {
+        (*OSSL_EC_POINT_free)(generator);
+        (*OSSL_EC_GROUP_free)(group);
+        (*OSSL_EC_KEY_free)(key);
+        return -1;
+    }
+
+    return (jlong)(intptr_t)key;
+}
+
+/* Encode an EC Elliptic Curve over a Binary Field
+ *
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    ECEncodeGF2m
+ * Signature: ([BI[BI[BI[BI[BI[BI[BI)J
+ */
+JNIEXPORT jlong JNICALL
+Java_jdk_crypto_jniprovider_NativeCrypto_ECEncodeGF2m
+  (JNIEnv *env, jclass obj, jbyteArray a, jint aLen, jbyteArray b, jint bLen, jbyteArray p, jint pLen, jbyteArray x, jint xLen, jbyteArray y, jint yLen, jbyteArray n, jint nLen, jbyteArray h, jint hLen)
+{
+    if (OSSL_NO_EC2M) {
+        return -1;
+    }
```

Comment:
This is not legal C code; the `if` statement must occur after the declarations in this block.